File: objectfabric/gwt/src/main/java/java/util/concurrent/ConcurrentHashMap.java
/**
 * This file is part of ObjectFabric (http://objectfabric.org).
 *
 * ObjectFabric is licensed under the Apache License, Version 2.0, the terms
 * of which may be found at http://www.apache.org/licenses/LICENSE-2.0.html.
 * 
 * Copyright ObjectFabric Inc.
 * 
 * This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
 * WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */

package java.util.concurrent;

import java.util.HashMap;

public class ConcurrentHashMap<K, V> extends HashMap<K, V> {

    public V putIfAbsent(K key, V value) {
        if (!containsKey(key))
            return put(key, value);

        return get(key);
    }

    public boolean replace(K key, V oldValue, V newValue) {
        if (containsKey(key) && get(key).equals(oldValue)) {
            put(key, newValue);
            return true;
        }

        return false;
    }

    public boolean remove(Object key, Object value) {
        if (containsKey(key) && get(key).equals(value)) {
            remove(key);
            return true;
        }

        return false;
    }
}

File: objectfabric/clr/Shared1/ConcurrentHashMap.cs
///
/// This file is part of ObjectFabric (http://objectfabric.org).
///
/// ObjectFabric is licensed under the Apache License, Version 2.0, the terms
/// of which may be found at http://www.apache.org/licenses/LICENSE-2.0.html.
///
/// Copyright ObjectFabric Inc.
///
/// This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
/// WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
///

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;

namespace ObjectFabric
{
    public class ConcurrentHashMap
    {
        // Cannot compile when class derives from this
        private readonly ConcurrentDictionary<object, object> _impl = new ConcurrentDictionary<object, object>();

        protected void Clear()
        {
            _impl.Clear();
        }

        protected bool ContainsKey(object key)
        {
            return _impl.ContainsKey(key);
        }

        protected java.lang.Iterable EntrySet()
        {
            return new EntrySet(_impl);
        }

        protected object Get(object key)
        {
            object value;

            if (_impl.TryGetValue(key, out value))
                return value;

            return null;
        }

        protected bool IsEmpty()
        {
            return _impl.IsEmpty;
        }

        public java.lang.Iterable KeySet()
        {
            return new EnumerableIterable(_impl.Keys);
        }

        protected object Put(object key, object value)
        {
            object previous = null;

            _impl.AddOrUpdate(key, value,
                (Func<object, object, object>) delegate(object oldKey, object oldValue)
                {
                    previous = oldValue;
                    return value;
                });

            return previous;
        }

        protected object PutIfAbsent(object key, object value)
        {
            object previous = null;
            _impl.AddOrUpdate(key, value, (oldKey, oldValue) => previous = oldValue);
            return previous;
        }

        protected object Remove(object key)
        {
            object value;
            _impl.TryRemove(key, out value);
            return value;
        }

        protected bool Remove(object key, object expect)
        {
            return ((ICollection<KeyValuePair<object, object>>) _impl).Remove(new KeyValuePair<object, object>(key, expect));
        }

        protected bool Replace(object key, object expect, object update)
        {
            return _impl.TryUpdate(key, update, expect);
        }

        protected int Size()
        {
            return _impl.Count;
        }

        public java.lang.Iterable Values()
        {
            return new EnumerableIterable(_impl.Values);
        }
    }
}
